[FREEZE] ERROR at teardown of TestsProcessPoolForkserverExecutor.test_map_chunksize

(randomly) observed on the CI on master:

- `test_loky linux-py37`

```
2022-02-03T15:49:22.5326916Z tests/test_process_executor_forkserver.py::TestsProcessPoolForkserverExecutor::test_map_chunksize PASSED [ 30%]
2022-02-03T15:49:22.5337930Z tests/test_process_executor_forkserver.py::TestsProcessPoolForkserverExecutor::test_map_chunksize ERROR [ 30%]
2022-02-03T15:49:34.3116493Z tests/test_process_executor_forkserver.py::TestsProcessPoolForkserverExecutor::test_traceback ERROR [ 31%]
2022-02-03T15:49:34.3127891Z tests/test_process_executor_forkserver.py::TestsProcessPoolForkserverExecutor::test_traceback ERROR [ 31%]
2022-02-03T15:49:46.1632822Z tests/test_process_executor_forkserver.py::TestsProcessPoolForkserverExecutor::test_thread_safety ERROR [ 31%]
2022-02-03T15:49:46.1691274Z tests/test_process_executor_forkserver.py::TestsProcessPoolForkserverExecutor::test_thread_safety ERROR [ 31%]teardown class with <multiprocessing.context.ForkServerContext object at 0x7f2f65670650>
2022-02-03T15:49:46.1693555Z 
2022-02-03T15:49:46.1694429Z 
2022-02-03T15:49:46.1695774Z ==================================== ERRORS ====================================
2022-02-03T15:49:46.1697429Z __ ERROR at teardown of TestsProcessPoolForkserverExecutor.test_map_chunksize __
2022-02-03T15:49:46.1699014Z 
2022-02-03T15:49:46.1700462Z self = <tests.test_process_executor_forkserver.TestsProcessPoolForkserverExecutor object at 0x7f2f5c869750>
2022-02-03T15:49:46.1707111Z method = <bound method ExecutorTest.test_map_chunksize of <tests.test_process_executor_forkserver.TestsProcessPoolForkserverExecutor object at 0x7f2f5c869750>>
2022-02-03T15:49:46.1707839Z 
2022-02-03T15:49:46.1708224Z     def teardown_method(self, method):
2022-02-03T15:49:46.1708679Z         # Make sure executor is not broken if it should not be
2022-02-03T15:49:46.1710048Z         executor = getattr(self, 'executor', None)
2022-02-03T15:49:46.1710502Z         if executor is not None:
2022-02-03T15:49:46.1710974Z             expect_broken_pool = hasattr(method, "broken_pool")  # old pytest
2022-02-03T15:49:46.1711477Z             for mark in getattr(method, "pytestmark", []):
2022-02-03T15:49:46.1711927Z                 if mark.name == "broken_pool":
2022-02-03T15:49:46.1712342Z                     expect_broken_pool = True
2022-02-03T15:49:46.1712816Z             is_actually_broken = executor._flags.broken is not None
2022-02-03T15:49:46.1713315Z             assert is_actually_broken == expect_broken_pool
2022-02-03T15:49:46.1713706Z     
2022-02-03T15:49:46.1714039Z             t_start = time.time()
2022-02-03T15:49:46.1714479Z >           executor.shutdown(wait=True, kill_workers=True)
2022-02-03T15:49:46.1714729Z 
2022-02-03T15:49:46.1715173Z executor   = <loky.process_executor.ProcessPoolExecutor object at 0x7f2f5c876290>
2022-02-03T15:49:46.1715673Z expect_broken_pool = False
2022-02-03T15:49:46.1716065Z is_actually_broken = False
2022-02-03T15:49:46.1716736Z method     = <bound method ExecutorTest.test_map_chunksize of <tests.test_process_executor_forkserver.TestsProcessPoolForkserverExecutor object at 0x7f2f5c869750>>
2022-02-03T15:49:46.1717595Z self       = <tests.test_process_executor_forkserver.TestsProcessPoolForkserverExecutor object at 0x7f2f5c869750>
2022-02-03T15:49:46.1718168Z t_start    = 1643903302.785902
2022-02-03T15:49:46.1718371Z 
2022-02-03T15:49:46.1718728Z tests/_executor_mixin.py:154: 
2022-02-03T15:49:46.1719181Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2022-02-03T15:49:46.1719682Z loky/process_executor.py:1157: in shutdown
2022-02-03T15:49:46.1720129Z     executor_manager_thread.join()
2022-02-03T15:49:46.1720707Z         executor_manager_thread = <_ExecutorManagerThread(ExecutorManagerThread, started daemon 139841316583168)>
2022-02-03T15:49:46.1721432Z         executor_manager_thread_wakeup = <loky.process_executor._ThreadWakeup object at 0x7f2f5c876c50>
2022-02-03T15:49:46.1721961Z         kill_workers = True
2022-02-03T15:49:46.1722453Z         self       = <loky.process_executor.ProcessPoolExecutor object at 0x7f2f5c876290>
2022-02-03T15:49:46.1722938Z         wait       = True
2022-02-03T15:49:46.1723421Z /opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/threading.py:1044: in join
2022-02-03T15:49:46.1723926Z     self._wait_for_tstate_lock()
2022-02-03T15:49:46.1724469Z         self       = <_ExecutorManagerThread(ExecutorManagerThread, started daemon 139841316583168)>
2022-02-03T15:49:46.1724979Z         timeout    = None
2022-02-03T15:49:46.1725406Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2022-02-03T15:49:46.1725660Z 
2022-02-03T15:49:46.1726122Z self = <_ExecutorManagerThread(ExecutorManagerThread, started daemon 139841316583168)>
2022-02-03T15:49:46.1726838Z block = True, timeout = -1
2022-02-03T15:49:46.1727043Z 
2022-02-03T15:49:46.1727647Z     def _wait_for_tstate_lock(self, block=True, timeout=-1):
2022-02-03T15:49:46.1728163Z         # Issue #18808: wait for the thread state to be gone.
2022-02-03T15:49:46.1728894Z         # At the end of the thread's life, after all knowledge of the thread
2022-02-03T15:49:46.1729471Z         # is removed from C data structures, C code releases our _tstate_lock.
2022-02-03T15:49:46.1730036Z         # This method passes its arguments to _tstate_lock.acquire().
2022-02-03T15:49:46.1730717Z         # If the lock is acquired, the C code is done, and self._stop() is
2022-02-03T15:49:46.1731255Z         # called.  That sets ._is_stopped to True, and ._tstate_lock to None.
2022-02-03T15:49:46.1731736Z         lock = self._tstate_lock
2022-02-03T15:49:46.1732196Z         if lock is None:  # already determined that the C code is done
2022-02-03T15:49:46.1732661Z             assert self._is_stopped
2022-02-03T15:49:46.1733059Z >       elif lock.acquire(block, timeout):
2022-02-03T15:49:46.1733552Z E       Failed: Timeout >60.0s
2022-02-03T15:49:46.1733755Z 
2022-02-03T15:49:46.1734044Z block      = True
2022-02-03T15:49:46.1734465Z lock       = <locked _thread.lock object at 0x7f2f5c880e10>
2022-02-03T15:49:46.1735046Z self       = <_ExecutorManagerThread(ExecutorManagerThread, started daemon 139841316583168)>
2022-02-03T15:49:46.1735714Z timeout    = -1
2022-02-03T15:49:46.1735892Z 
2022-02-03T15:49:46.1736344Z /opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/threading.py:1060: Failed
2022-02-03T15:49:46.1737159Z ---------------------------- Captured stderr setup -----------------------------
2022-02-03T15:49:46.1737784Z [DEBUG:MainProcess:MainThread] created semlock with handle 139841392947200
2022-02-03T15:49:46.1738405Z [DEBUG:MainProcess:MainThread] created semlock with handle 139841392943104
2022-02-03T15:49:46.1739018Z [DEBUG:MainProcess:MainThread] created semlock with handle 139841392939008
2022-02-03T15:49:46.1739630Z [DEBUG:MainProcess:MainThread] created semlock with handle 139841392934912
2022-02-03T15:49:46.1740183Z [DEBUG:MainProcess:MainThread] Queue._after_fork()
2022-02-03T15:49:46.1740746Z [DEBUG:MainProcess:MainThread] created semlock with handle 139841392930816
2022-02-03T15:49:46.1741360Z [DEBUG:MainProcess:MainThread] created semlock with handle 139841392926720
2022-02-03T15:49:46.1743391Z [DEBUG:MainProcess:MainThread] ProcessPoolExecutor is setup
2022-02-03T15:49:46.1744209Z [DEBUG:MainProcess:MainThread] created semlock with handle 139841392898048
2022-02-03T15:49:46.1744940Z [DEBUG:MainProcess:MainThread] created semlock with handle 139841392893952
2022-02-03T15:49:46.1745649Z [DEBUG:MainProcess:MainThread] created semlock with handle 139841392889856
2022-02-03T15:49:46.1746355Z [DEBUG:MainProcess:MainThread] created semlock with handle 139841392885760
2022-02-03T15:49:46.1747076Z [DEBUG:MainProcess:MainThread] created semlock with handle 139841392881664
2022-02-03T15:49:46.1749031Z [DEBUG:MainProcess:MainThread] Adjust process count : {2673: <ForkServerProcess(ForkServerProcess-207, started)>, 2674: <ForkServerProcess(ForkServerProcess-208, started)>, 2675: <ForkServerProcess(ForkServerProcess-209, started)>, 2676: <ForkServerProcess(ForkServerProcess-210, started)>, 2677: <ForkServerProcess(ForkServerProcess-211, started)>}
2022-02-03T15:49:46.1750426Z [DEBUG:MainProcess:MainThread] _start_executor_manager_thread called
2022-02-03T15:49:46.1751116Z [DEBUG:MainProcess:ExecutorManagerThread] Queue._start_thread()
2022-02-03T15:49:46.1751802Z [DEBUG:MainProcess:ExecutorManagerThread] doing self._thread.start()
2022-02-03T15:49:46.1752503Z [DEBUG:MainProcess:QueueFeederThread] starting thread to feed data to pipe
2022-02-03T15:49:46.1753202Z [DEBUG:MainProcess:ExecutorManagerThread] ... done self._thread.start()
2022-02-03T15:49:46.1754104Z [DEBUG/ForkServerProcess-207] Using 'cloudpickle' for serialization.
2022-02-03T15:49:46.1754975Z [DEBUG/ForkServerProcess-211] Using 'cloudpickle' for serialization.
2022-02-03T15:49:46.1755889Z [DEBUG/ForkServerProcess-207] recreated blocker with handle 139841392943104
2022-02-03T15:49:46.1822137Z [DEBUG/ForkServerProcess-207] recreated blocker with handle 139841392939008
2022-02-03T15:49:46.1823521Z [DEBUG/ForkServerProcess-207] recreated blocker with handle 139841392934912
2022-02-03T15:49:46.1824413Z [DEBUG/ForkServerProcess-207] Queue._after_fork()
2022-02-03T15:49:46.1825314Z [DEBUG/ForkServerProcess-207] recreated blocker with handle 139841392930816
2022-02-03T15:49:46.1826250Z [DEBUG/ForkServerProcess-207] recreated blocker with handle 139841392926720
2022-02-03T15:49:46.1827941Z [DEBUG/ForkServerProcess-207] recreated blocker with handle 139841541672960
2022-02-03T15:49:46.1828917Z [DEBUG/ForkServerProcess-207] recreated blocker with handle 139841541668864
2022-02-03T15:49:46.1829977Z [DEBUG/ForkServerProcess-209] Using 'cloudpickle' for serialization.
2022-02-03T15:49:46.1830924Z [DEBUG/ForkServerProcess-207] recreated blocker with handle 139841541664768
2022-02-03T15:49:46.1831858Z [DEBUG/ForkServerProcess-207] recreated blocker with handle 139841541660672
2022-02-03T15:49:46.1833019Z [DEBUG/ForkServerProcess-207] recreated blocker with handle 139841541607424
2022-02-03T15:49:46.1834749Z [DEBUG/ForkServerProcess-207] recreated blocker with handle 139841392947200
2022-02-03T15:49:46.1838919Z [DEBUG/ForkServerProcess-207] recreated blocker with handle 139841392898048
2022-02-03T15:49:46.1840247Z [INFO/ForkServerProcess-207] child process calling self.run()
2022-02-03T15:49:46.1841155Z [DEBUG/ForkServerProcess-207] Worker started with timeout=None
2022-02-03T15:49:46.1842085Z [DEBUG/ForkServerProcess-211] recreated blocker with handle 139841392943104
2022-02-03T15:49:46.1843006Z [DEBUG/ForkServerProcess-211] recreated blocker with handle 139841392939008
2022-02-03T15:49:46.1843949Z [DEBUG/ForkServerProcess-211] recreated blocker with handle 139841392934912
2022-02-03T15:49:46.1844812Z [DEBUG/ForkServerProcess-211] Queue._after_fork()
2022-02-03T15:49:46.1845688Z [DEBUG/ForkServerProcess-211] recreated blocker with handle 139841392930816
2022-02-03T15:49:46.1846604Z [DEBUG/ForkServerProcess-211] recreated blocker with handle 139841392926720
2022-02-03T15:49:46.1847545Z ----------------------------- Captured stderr call -----------------------------
2022-02-03T15:49:46.1848457Z [DEBUG/ForkServerProcess-208] Using 'cloudpickle' for serialization.
2022-02-03T15:49:46.1849374Z [DEBUG/ForkServerProcess-211] recreated blocker with handle 139841541672960
2022-02-03T15:49:46.1850303Z [DEBUG/ForkServerProcess-211] recreated blocker with handle 139841541668864
2022-02-03T15:49:46.1851181Z [DEBUG/ForkServerProcess-211] recreated blocker with handle 139841541664768
2022-02-03T15:49:46.1852058Z [DEBUG/ForkServerProcess-211] recreated blocker with handle 139841541660672
2022-02-03T15:49:46.1852934Z [DEBUG/ForkServerProcess-211] recreated blocker with handle 139841541607424
2022-02-03T15:49:46.1853804Z [DEBUG/ForkServerProcess-211] recreated blocker with handle 139841392947200
2022-02-03T15:49:46.1854674Z [DEBUG/ForkServerProcess-211] recreated blocker with handle 139841392881664
2022-02-03T15:49:46.1855494Z [INFO/ForkServerProcess-211] child process calling self.run()
2022-02-03T15:49:46.1856297Z [DEBUG/ForkServerProcess-211] Worker started with timeout=None
2022-02-03T15:49:46.1857109Z [DEBUG/ForkServerProcess-207] psutil return memory size: 25661440
2022-02-03T15:49:46.1857963Z [DEBUG/ForkServerProcess-209] recreated blocker with handle 139841392943104
2022-02-03T15:49:46.1858822Z [DEBUG/ForkServerProcess-209] recreated blocker with handle 139841392939008
2022-02-03T15:49:46.1859699Z [DEBUG/ForkServerProcess-209] recreated blocker with handle 139841392934912
2022-02-03T15:49:46.1860489Z [DEBUG/ForkServerProcess-209] Queue._after_fork()
2022-02-03T15:49:46.1861310Z [DEBUG/ForkServerProcess-209] recreated blocker with handle 139841392930816
2022-02-03T15:49:46.1862375Z [DEBUG/ForkServerProcess-209] recreated blocker with handle 139841392926720
2022-02-03T15:49:46.1863263Z [DEBUG/ForkServerProcess-209] recreated blocker with handle 139841541672960
2022-02-03T15:49:46.1864137Z [DEBUG/ForkServerProcess-209] recreated blocker with handle 139841541668864
2022-02-03T15:49:46.1865012Z [DEBUG/ForkServerProcess-209] recreated blocker with handle 139841541664768
2022-02-03T15:49:46.1865889Z [DEBUG/ForkServerProcess-209] recreated blocker with handle 139841541660672
2022-02-03T15:49:46.1866750Z [DEBUG/ForkServerProcess-209] recreated blocker with handle 139841541607424
2022-02-03T15:49:46.1867620Z [DEBUG/ForkServerProcess-209] recreated blocker with handle 139841392947200
2022-02-03T15:49:46.1868693Z [DEBUG/ForkServerProcess-209] recreated blocker with handle 139841392889856
2022-02-03T15:49:46.1869518Z [INFO/ForkServerProcess-209] child process calling self.run()
2022-02-03T15:49:46.1870318Z [DEBUG/ForkServerProcess-209] Worker started with timeout=None
2022-02-03T15:49:46.1871113Z [DEBUG/ForkServerProcess-211] psutil return memory size: 25665536
2022-02-03T15:49:46.1872242Z [DEBUG/ForkServerProcess-210] Using 'cloudpickle' for serialization.
2022-02-03T15:49:46.1873132Z --------------------------- Captured stderr teardown ---------------------------
2022-02-03T15:49:46.1874066Z [DEBUG:MainProcess:MainThread] shutting down executor <loky.process_executor.ProcessPoolExecutor object at 0x7f2f5c876290>
2022-02-03T15:49:46.1875211Z [DEBUG:MainProcess:ExecutorManagerThread] terminate process ForkServerProcess-211, reason: executor shutting down
2022-02-03T15:49:46.1876179Z [DEBUG/ForkServerProcess-208] recreated blocker with handle 139841392943104
2022-02-03T15:49:46.1877061Z [DEBUG/ForkServerProcess-208] recreated blocker with handle 139841392939008
2022-02-03T15:49:46.1877930Z [DEBUG/ForkServerProcess-208] recreated blocker with handle 139841392934912
2022-02-03T15:49:46.1878728Z [DEBUG/ForkServerProcess-208] Queue._after_fork()
2022-02-03T15:49:46.1879534Z [DEBUG/ForkServerProcess-208] recreated blocker with handle 139841392930816
2022-02-03T15:49:46.1880409Z [DEBUG/ForkServerProcess-208] recreated blocker with handle 139841392926720
2022-02-03T15:49:46.1881278Z [DEBUG/ForkServerProcess-208] recreated blocker with handle 139841541672960
2022-02-03T15:49:46.1882154Z [DEBUG/ForkServerProcess-208] recreated blocker with handle 139841541668864
2022-02-03T15:49:46.1883024Z [DEBUG/ForkServerProcess-208] recreated blocker with handle 139841541664768
2022-02-03T15:49:46.1883873Z [DEBUG/ForkServerProcess-208] recreated blocker with handle 139841541660672
2022-02-03T15:49:46.1884742Z [DEBUG/ForkServerProcess-208] recreated blocker with handle 139841541607424
2022-02-03T15:49:46.1885618Z [DEBUG/ForkServerProcess-208] recreated blocker with handle 139841392947200
2022-02-03T15:49:46.1886456Z [DEBUG/ForkServerProcess-209] psutil return memory size: 25669632
2022-02-03T15:49:46.1887304Z [DEBUG/ForkServerProcess-208] recreated blocker with handle 139841392893952
2022-02-03T15:49:46.1888110Z [INFO/ForkServerProcess-208] child process calling self.run()
2022-02-03T15:49:46.1888905Z [DEBUG/ForkServerProcess-208] Worker started with timeout=None
2022-02-03T15:49:46.1889747Z [DEBUG/ForkServerProcess-210] recreated blocker with handle 139841392943104
2022-02-03T15:49:46.1890619Z [DEBUG/ForkServerProcess-210] recreated blocker with handle 139841392939008
2022-02-03T15:49:46.1891473Z [DEBUG/ForkServerProcess-210] recreated blocker with handle 139841392934912
2022-02-03T15:49:46.1892264Z [DEBUG/ForkServerProcess-210] Queue._after_fork()
2022-02-03T15:49:46.1893080Z [DEBUG/ForkServerProcess-210] recreated blocker with handle 139841392930816
2022-02-03T15:49:46.1893947Z [DEBUG/ForkServerProcess-210] recreated blocker with handle 139841392926720
2022-02-03T15:49:46.1894817Z [DEBUG/ForkServerProcess-210] recreated blocker with handle 139841541672960
2022-02-03T15:49:46.1895671Z [DEBUG/ForkServerProcess-210] recreated blocker with handle 139841541668864
2022-02-03T15:49:46.1907277Z [DEBUG/ForkServerProcess-210] recreated blocker with handle 139841541664768
2022-02-03T15:49:46.1908377Z [DEBUG/ForkServerProcess-210] recreated blocker with handle 139841541660672
2022-02-03T15:49:46.1909266Z [DEBUG/ForkServerProcess-210] recreated blocker with handle 139841541607424
2022-02-03T15:49:46.1910137Z [DEBUG/ForkServerProcess-210] recreated blocker with handle 139841392947200
2022-02-03T15:49:46.1911012Z [DEBUG/ForkServerProcess-210] recreated blocker with handle 139841392885760
2022-02-03T15:49:46.1911844Z [INFO/ForkServerProcess-210] child process calling self.run()
2022-02-03T15:49:46.1912647Z [DEBUG/ForkServerProcess-210] Worker started with timeout=None
2022-02-03T15:49:46.1912985Z 
2022-02-03T15:49:46.1913672Z +++++++++++++++++++++++++++++++++++ Timeout ++++++++++++++++++++++++++++++++++++
2022-02-03T15:49:46.1914022Z 
2022-02-03T15:49:46.1914512Z ~~~~~~~~~~~~~~~~~ Stack of QueueFeederThread (139841299801856) ~~~~~~~~~~~~~~~~~
2022-02-03T15:49:46.1915212Z   File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/threading.py", line 890, in _bootstrap
2022-02-03T15:49:46.1915808Z     self._bootstrap_inner()
2022-02-03T15:49:46.1916388Z   File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/threading.py", line 926, in _bootstrap_inner
2022-02-03T15:49:46.1917072Z     self.run()
2022-02-03T15:49:46.1917611Z   File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/threading.py", line 870, in run
2022-02-03T15:49:46.1918238Z     self._target(*self._args, **self._kwargs)
2022-02-03T15:49:46.1918778Z   File "/home/vsts/work/1/s/loky/backend/queues.py", line 141, in _feed
2022-02-03T15:49:46.1919271Z     nwait()
2022-02-03T15:49:46.1919800Z   File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/threading.py", line 296, in wait
2022-02-03T15:49:46.1920372Z     waiter.acquire()
2022-02-03T15:49:46.1920596Z 
2022-02-03T15:49:46.1921080Z ~~~~~~~~~~~~~~~ Stack of ExecutorManagerThread (139841316583168) ~~~~~~~~~~~~~~~
2022-02-03T15:49:46.1921776Z   File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/threading.py", line 890, in _bootstrap
2022-02-03T15:49:46.1922368Z     self._bootstrap_inner()
2022-02-03T15:49:46.1922963Z   File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/threading.py", line 926, in _bootstrap_inner
2022-02-03T15:49:46.1923524Z     self.run()
2022-02-03T15:49:46.1924011Z   File "/home/vsts/work/1/s/loky/process_executor.py", line 558, in run
2022-02-03T15:49:46.1924558Z     self.flag_executor_shutting_down()
2022-02-03T15:49:46.1925156Z   File "/home/vsts/work/1/s/loky/process_executor.py", line 754, in flag_executor_shutting_down
2022-02-03T15:49:46.1925772Z     self.kill_workers(reason="executor shutting down")
2022-02-03T15:49:46.1926369Z   File "/home/vsts/work/1/s/loky/process_executor.py", line 764, in kill_workers
2022-02-03T15:49:46.1926917Z     recursive_terminate(p)
2022-02-03T15:49:46.1927454Z   File "/home/vsts/work/1/s/loky/backend/utils.py", line 26, in recursive_terminate
2022-02-03T15:49:46.1928044Z     _recursive_terminate_with_psutil(process)
2022-02-03T15:49:46.1928642Z   File "/home/vsts/work/1/s/loky/backend/utils.py", line 46, in _recursive_terminate_with_psutil
2022-02-03T15:49:46.1929203Z     process.join()
2022-02-03T15:49:46.1929783Z   File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/multiprocessing/process.py", line 140, in join
2022-02-03T15:49:46.1930401Z     res = self._popen.wait(timeout)
2022-02-03T15:49:46.1931721Z   File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/multiprocessing/popen_fork.py", line 48, in wait
2022-02-03T15:49:46.1933220Z     return self.poll(os.WNOHANG if timeout == 0.0 else 0)
2022-02-03T15:49:46.1933953Z   File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/multiprocessing/popen_forkserver.py", line 61, in poll
2022-02-03T15:49:46.1934629Z     if not wait([self.sentinel], timeout):
2022-02-03T15:49:46.1935258Z   File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/multiprocessing/connection.py", line 921, in wait
2022-02-03T15:49:46.1935874Z     ready = selector.select(timeout)
2022-02-03T15:49:46.1936463Z   File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/selectors.py", line 415, in select
2022-02-03T15:49:46.1941371Z     fd_event_list = self._selector.poll(timeout)
2022-02-03T15:49:46.1941719Z 
2022-02-03T15:49:46.1942993Z ~~~~~~~~~~~~~~~~~~~~~ Stack of Thread-12 (139841485211392) ~~~~~~~~~~~~~~~~~~~~~
2022-02-03T15:49:46.1943738Z   File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/threading.py", line 890, in _bootstrap
2022-02-03T15:49:46.1944334Z     self._bootstrap_inner()
2022-02-03T15:49:46.1944918Z   File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/threading.py", line 926, in _bootstrap_inner
2022-02-03T15:49:46.1945490Z     self.run()
2022-02-03T15:49:46.1946287Z   File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/threading.py", line 870, in run
2022-02-03T15:49:46.1946888Z     self._target(*self._args, **self._kwargs)
2022-02-03T15:49:46.1947528Z   File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/multiprocessing/resource_sharer.py", line 142, in _serve
2022-02-03T15:49:46.1948182Z     with self._listener.accept() as conn:
2022-02-03T15:49:46.1948818Z   File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/multiprocessing/connection.py", line 453, in accept
2022-02-03T15:49:46.1949554Z     c = self._listener.accept()
2022-02-03T15:49:46.1950165Z   File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/multiprocessing/connection.py", line 599, in accept
2022-02-03T15:49:46.1950802Z     s, self._last_accepted = self._socket.accept()
2022-02-03T15:49:46.1951414Z   File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/socket.py", line 212, in accept
2022-02-03T15:49:46.1951984Z     fd, addr = self._accept()
2022-02-03T15:49:46.1952221Z 
2022-02-03T15:49:46.1952711Z +++++++++++++++++++++++++++++++++++ Timeout ++++++++++++++++++++++++++++++++++++
2022-02-03T15:49:46.1953365Z _____ ERROR at setup of TestsProcessPoolForkserverExecutor.test_traceback ______
```

https://dev.azure.com/joblib/loky/_build/results?buildId=1574&view=logs&j=690d157a-d5ad-54bf-a85d-89b5d05cf663&t=35eaaa5a-db59-5698-331c-8016de4f39a3

Similar problems:

-   #346 `ERROR at teardown of TestsProcessPoolLokyExecutor.test_killed_child`:
    https://dev.azure.com/joblib/loky/_build/results?buildId=1751&view=logs&j=0bee2fdb-ff38-51c4-8e0e-ffcd1c318aed&t=9bcb10dc-bb71-5de7-d670-7b816012a9d5

```python-traceback
_____ ERROR at teardown of TestsProcessPoolLokyExecutor.test_killed_child ______

self = <tests.test_process_executor_loky.TestsProcessPoolLokyExecutor object at 0x7fd459298a60>
method = <bound method ExecutorTest.test_killed_child of <tests.test_process_executor_loky.TestsProcessPoolLokyExecutor object at 0x7fd459298a60>>

    def teardown_method(self, method):
        # Make sure executor is not broken if it should not be
        executor = getattr(self, 'executor', None)
        if executor is not None:
            expect_broken_pool = hasattr(method, "broken_pool")  # old pytest
            for mark in getattr(method, "pytestmark", []):
                if mark.name == "broken_pool":
                    expect_broken_pool = True
            is_actually_broken = executor._flags.broken is not None
>           assert is_actually_broken == expect_broken_pool
E           AssertionError

executor   = <loky.process_executor.ProcessPoolExecutor object at 0x7fd456a8efb0>
expect_broken_pool = True
is_actually_broken = False
mark       = Mark(name='broken_pool', args=(), kwargs={})
method     = <bound method ExecutorTest.test_killed_child of <tests.test_process_executor_loky.TestsProcessPoolLokyExecutor object at 0x7fd459298a60>>
self       = <tests.test_process_executor_loky.TestsProcessPoolLokyExecutor object at 0x7fd459298a60>
```

--- a/.azure_pipeline.yml
+++ b/.azure_pipeline.yml
@@ -9,8 +9,8 @@ jobs:
         tox.env: py310
       windows-py37:
         imageName: windows-latest
-        python.version: "3.7"
-        tox.env: py37
+        python.version: "3.8"
+        tox.env: py38
 
       macos-py310:
         imageName: "macos-10.15"
@@ -36,15 +36,15 @@ jobs:
         python.version: "3.9"
         tox.env: "py39"
         joblib.tests: "true"
-      linux-python-py38-high-memory:
+      linux-python-py39-high-memory:
+        imageName: "ubuntu-20.04"
+        python.version: "3.9"
+        tox.env: py37
+        RUN_MEMORY: "true"
+      linux-py38:
         imageName: "ubuntu-20.04"
         python.version: "3.8"
         tox.env: py38
-        RUN_MEMORY: "true"
-      linux-py37:
-        imageName: "ubuntu-20.04"
-        python.version: "3.7"
-        tox.env: py37
 
   pool:
     vmImage: $(imageName)

--- a/.azure_pipeline.yml
+++ b/.azure_pipeline.yml
@@ -39,7 +39,7 @@ jobs:
       linux-python-py39-high-memory:
         imageName: "ubuntu-20.04"
         python.version: "3.9"
-        tox.env: py37
+        tox.env: py39
         RUN_MEMORY: "true"
       linux-py38:
         imageName: "ubuntu-20.04"

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -4,6 +4,9 @@
   garbage collecting an executor instance when shutting down the
   the Python interpreter (#311).
 
+- Make `shutdown(kill_workers=True)` consistently use the SIGKILL
+  signal on POSIX. Previously a mix of SIGKILL and SIGTERM was issued
+  and could deadlock the shutdown process (#348).
 
 ### 3.0.0 - 2021-09-10
 

--- a/loky/backend/utils.py
+++ b/loky/backend/utils.py
@@ -74,14 +74,14 @@ def _kill_process_tree_without_psutil(process):
                 raise
     else:
         try:
-            _kill_process_tree(process.pid)
-        except OSError as e:
+            _recursive_kill_posix(process.pid)
+        except OSError:
             warnings.warn(
                 "Failed to kill subprocesses on this platform. Please install"
                 "psutil: https://github.com/giampaolo/psutil"
             )
             # In case we cannot introspect the children, we fall back to only
-            # kill the main process.
+            # killing the main process.
             process.kill()
     process.join()
 
@@ -102,7 +102,7 @@ def _kill(pid):
             raise
 
 
-def _kill_process_tree(pid):
+def _recursive_kill_posix(pid):
     """Recursively kill the descendants of a process before killing it."""
     try:
         children_pids = subprocess.check_output(
@@ -118,7 +118,7 @@ def _kill_process_tree(pid):
     # Decode the result, split the cpid and remove the trailing line
     for cpid in children_pids.splitlines():
         cpid = int(cpid)
-        _kill_process_tree(cpid)
+        _recursive_kill_posix(cpid)
 
     _kill(pid)
 

--- a/loky/backend/utils.py
+++ b/loky/backend/utils.py
@@ -13,7 +13,7 @@ except ImportError:
 
 def _kill(pid):
     if sys.platform == "win32":
-        kill_signal = signal.CTRL_BREAK_EVENT
+        kill_signal = signal.SIGTERM
     else:
         kill_signal = signal.SIGKILL
     try:

--- a/loky/backend/utils.py
+++ b/loky/backend/utils.py
@@ -11,14 +11,39 @@ except ImportError:
     psutil = None
 
 
-def recursive_terminate(process, use_psutil=True):
-    if use_psutil and psutil is not None:
-        _recursive_terminate_with_psutil(process)
+def _kill(pid):
+    if sys.platform == "win32":
+        kill_signal = signal.CTRL_BREAK_EVENT
     else:
-        _recursive_terminate_without_psutil(process)
+        kill_signal = signal.SIGKILL
+    try:
+        os.kill(pid, kill_signal)
+    except OSError as e:
+        # if OSError is raised with [Errno 3] no such process, the process
+        # is already terminated, else, raise the error and let the top
+        # level function raise a warning and retry to kill the process.
+        if e.errno != errno.ESRCH:
+            raise
 
 
-def _recursive_terminate_with_psutil(process):
+def recursive_kill(process, use_psutil=True):
+    """Terminate process and its descendants with SIGKILL"""
+    if use_psutil and psutil is not None:
+        _recursive_kill_with_psutil(process)
+    else:
+        _recursive_kill_without_psutil(process)
+
+
+def recursive_terminate(process, use_psutil=True):
+    warnings.warn(
+        "recursive_terminate is deprecated in loky 3.2, use recursive_kill"
+        "instead",
+        DeprecationWarning,
+    )
+    recursive_kill(process, use_psutil=use_psutil)
+
+
+def _recursive_kill_with_psutil(process):
     try:
         children = psutil.Process(process.pid).children(recursive=True)
     except psutil.NoSuchProcess:
@@ -32,56 +57,51 @@ def _recursive_terminate_with_psutil(process):
         except psutil.NoSuchProcess:
             pass
 
-    process.terminate()
+    _kill(process.pid)
     process.join()
 
 
-def _recursive_terminate_without_psutil(process):
-    """Terminate a process and its descendants.
-    """
+def _recursive_kill_without_psutil(process):
+    """Terminate a process and its descendants."""
     try:
-        _recursive_terminate(process.pid)
+        _recursive_kill(process.pid)
     except OSError as e:
         warnings.warn("Failed to kill subprocesses on this platform. Please"
                       "install psutil: https://github.com/giampaolo/psutil")
-        # In case we cannot introspect the children, we fall back to the
-        # classic Process.terminate.
-        process.terminate()
+        # In case we cannot introspect the children, we fall back to only
+        # kill the main process.
+        _kill(process.pid)
     process.join()
 
 
-def _recursive_terminate(pid):
-    """Recursively kill the descendants of a process before killing it.
-    """
+def _recursive_kill(pid):
+    """Recursively kill the descendants of a process before killing it."""
 
     if sys.platform == "win32":
         # On windows, the taskkill function with option `/T` terminate a given
         # process pid and its children.
         try:
             subprocess.check_output(
-                ["taskkill", "/F", "/T", "/PID", str(pid)],
-                stderr=None)
+                ["taskkill", "/F", "/T", "/PID", str(pid)], stderr=None
+            )
         except subprocess.CalledProcessError as e:
-            # In windows, taskkill return 1 for permission denied and 128, 255
+            # In Windows, taskkill returns 1 for permission denied and 128, 255
             # for no process found.
-            if e.returncode not in [1, 128, 255]:
-                raise
-            elif e.returncode == 1:
+            if e.returncode == 1:
                 # Try to kill the process without its descendants if taskkill
                 # was denied permission. If this fails too, with an error
                 # different from process not found, let the top level function
                 # raise a warning and retry to kill the process.
-                try:
-                    os.kill(pid, signal.SIGTERM)
-                except OSError as e:
-                    if e.errno != errno.ESRCH:
-                        raise
-
+                _kill(pid)
+            elif e.returncode in [128, 255]:
+                # Process not found: ignore.
+                return
+            else:
+                raise
     else:
         try:
             children_pids = subprocess.check_output(
-                ["pgrep", "-P", str(pid)],
-                stderr=None, text=True
+                ["pgrep", "-P", str(pid)], stderr=None, text=True
             )
         except subprocess.CalledProcessError as e:
             # `ps` returns 1 when no child process has been found
@@ -93,16 +113,9 @@ def _recursive_terminate(pid):
         # Decode the result, split the cpid and remove the trailing line
         for cpid in children_pids.splitlines():
             cpid = int(cpid)
-            _recursive_terminate(cpid)
+            _recursive_kill(cpid)
 
-        try:
-            os.kill(pid, signal.SIGTERM)
-        except OSError as e:
-            # if OSError is raised with [Errno 3] no such process, the process
-            # is already terminated, else, raise the error and let the top
-            # level function raise a warning and retry to kill the process.
-            if e.errno != errno.ESRCH:
-                raise
+        _kill(pid)
 
 
 def get_exitcodes_terminated_worker(processes):

--- a/loky/backend/utils.py
+++ b/loky/backend/utils.py
@@ -11,9 +11,86 @@ except ImportError:
     psutil = None
 
 
+def kill_process_tree(process, use_psutil=True):
+    """Terminate process and its descendants with SIGKILL"""
+    if use_psutil and psutil is not None:
+        _kill_process_tree_with_psutil(process)
+    else:
+        _kill_process_tree_without_psutil(process)
+
+
+def recursive_terminate(process, use_psutil=True):
+    warnings.warn(
+        "recursive_terminate is deprecated in loky 3.2, use kill_process_tree"
+        "instead",
+        DeprecationWarning,
+    )
+    kill_process_tree(process, use_psutil=use_psutil)
+
+
+def _kill_process_tree_with_psutil(process):
+    try:
+        descendants = psutil.Process(process.pid).children(recursive=True)
+    except psutil.NoSuchProcess:
+        return
+
+    # Kill the descendants in reverse order to avoid killing the parents before
+    # the descendant in cases where there are more processes nested.
+    for descendant in descendants[::-1]:
+        try:
+            descendant.kill()
+        except psutil.NoSuchProcess:
+            pass
+
+    try:
+        psutil.Process(process.pid).kill()
+    except psutil.NoSuchProcess:
+        pass
+    process.join()
+
+
+def _kill_process_tree_without_psutil(process):
+    """Terminate a process and its descendants."""
+    if sys.platform == "win32":
+        # On windows, the taskkill function with option `/T` terminate a given
+        # process pid and its children.
+        try:
+            subprocess.check_output(
+                ["taskkill", "/F", "/T", "/PID", str(process.pid)], stderr=None
+            )
+        except subprocess.CalledProcessError as e:
+            # In Windows, taskkill returns 1 for permission denied and 128, 255
+            # for no process found.
+            if e.returncode == 1:
+                # Try to terminate the process without its descendants if
+                # taskkill was denied permission. If this fails too, with an
+                # error different from process not found, let the top level
+                # function raise a warning and retry to kill the process.
+                process.terminate()
+            elif e.returncode in [128, 255]:
+                # Process not found: ignore.
+                pass
+            else:
+                raise
+    else:
+        try:
+            _kill_process_tree(process.pid)
+        except OSError as e:
+            warnings.warn(
+                "Failed to kill subprocesses on this platform. Please install"
+                "psutil: https://github.com/giampaolo/psutil"
+            )
+            # In case we cannot introspect the children, we fall back to only
+            # kill the main process.
+            process.kill()
+    process.join()
+
+
 def _kill(pid):
     # Not all systems (e.g. Windows) have a SIGKILL, but the C specification
-    # requires a SIGTERM signal.
+    # mandates a SIGTERM signal. While Windows is handled specifically above,
+    # let's try to be safe for other hypothetic platforms that only have
+    # SIGTERM without SIGKILL.
     kill_signal = getattr(signal, 'SIGKILL', signal.SIGTERM)
     try:
         os.kill(pid, kill_signal)
@@ -25,96 +102,25 @@ def _kill(pid):
             raise
 
 
-def recursive_kill(process, use_psutil=True):
-    """Terminate process and its descendants with SIGKILL"""
-    if use_psutil and psutil is not None:
-        _recursive_kill_with_psutil(process)
-    else:
-        _recursive_kill_without_psutil(process)
-
-
-def recursive_terminate(process, use_psutil=True):
-    warnings.warn(
-        "recursive_terminate is deprecated in loky 3.2, use recursive_kill"
-        "instead",
-        DeprecationWarning,
-    )
-    recursive_kill(process, use_psutil=use_psutil)
-
-
-def _recursive_kill_with_psutil(process):
-    try:
-        children = psutil.Process(process.pid).children(recursive=True)
-    except psutil.NoSuchProcess:
-        return
-
-    # Kill the children in reverse order to avoid killing the parents before
-    # the children in cases where there are more processes nested.
-    for child in children[::-1]:
-        try:
-            child.kill()
-        except psutil.NoSuchProcess:
-            pass
-
-    _kill(process.pid)
-    process.join()
-
-
-def _recursive_kill_without_psutil(process):
-    """Terminate a process and its descendants."""
-    try:
-        _recursive_kill(process.pid)
-    except OSError as e:
-        warnings.warn("Failed to kill subprocesses on this platform. Please"
-                      "install psutil: https://github.com/giampaolo/psutil")
-        # In case we cannot introspect the children, we fall back to only
-        # kill the main process.
-        _kill(process.pid)
-    process.join()
-
-
-def _recursive_kill(pid):
+def _kill_process_tree(pid):
     """Recursively kill the descendants of a process before killing it."""
+    try:
+        children_pids = subprocess.check_output(
+            ["pgrep", "-P", str(pid)], stderr=None, text=True
+        )
+    except subprocess.CalledProcessError as e:
+        # `ps` returns 1 when no child process has been found
+        if e.returncode == 1:
+            children_pids = ''
+        else:
+            raise
 
-    if sys.platform == "win32":
-        # On windows, the taskkill function with option `/T` terminate a given
-        # process pid and its children.
-        try:
-            subprocess.check_output(
-                ["taskkill", "/F", "/T", "/PID", str(pid)], stderr=None
-            )
-        except subprocess.CalledProcessError as e:
-            # In Windows, taskkill returns 1 for permission denied and 128, 255
-            # for no process found.
-            if e.returncode == 1:
-                # Try to kill the process without its descendants if taskkill
-                # was denied permission. If this fails too, with an error
-                # different from process not found, let the top level function
-                # raise a warning and retry to kill the process.
-                _kill(pid)
-            elif e.returncode in [128, 255]:
-                # Process not found: ignore.
-                return
-            else:
-                raise
-    else:
-        try:
-            children_pids = subprocess.check_output(
-                ["pgrep", "-P", str(pid)], stderr=None, text=True
-            )
-        except subprocess.CalledProcessError as e:
-            # `ps` returns 1 when no child process has been found
-            if e.returncode == 1:
-                children_pids = ''
-            else:
-                raise
+    # Decode the result, split the cpid and remove the trailing line
+    for cpid in children_pids.splitlines():
+        cpid = int(cpid)
+        _kill_process_tree(cpid)
 
-        # Decode the result, split the cpid and remove the trailing line
-        for cpid in children_pids.splitlines():
-            cpid = int(cpid)
-            _recursive_kill(cpid)
-
-        _kill(pid)
+    _kill(pid)
 
 
 def get_exitcodes_terminated_worker(processes):

--- a/loky/backend/utils.py
+++ b/loky/backend/utils.py
@@ -12,10 +12,9 @@ except ImportError:
 
 
 def _kill(pid):
-    if sys.platform == "win32":
-        kill_signal = signal.SIGTERM
-    else:
-        kill_signal = signal.SIGKILL
+    # Not all systems (e.g. Windows) have a SIGKILL, but the C specification
+    # requires a SIGTERM signal.
+    kill_signal = getattr(signal, 'SIGKILL', signal.SIGTERM)
     try:
         os.kill(pid, kill_signal)
     except OSError as e:

--- a/loky/process_executor.py
+++ b/loky/process_executor.py
@@ -84,7 +84,7 @@ from .backend import get_context
 from .backend.context import cpu_count
 from .backend.queues import Queue, SimpleQueue
 from .backend.reduction import set_loky_pickler, get_loky_pickler_name
-from .backend.utils import recursive_terminate, get_exitcodes_terminated_worker
+from .backend.utils import recursive_kill, get_exitcodes_terminated_worker
 from .initializers import _prepare_initializer
 
 
@@ -763,7 +763,7 @@ class _ExecutorManagerThread(threading.Thread):
             _, p = self.processes.popitem()
             mp.util.debug(f"terminate process {p.name}, reason: {reason}")
             try:
-                recursive_terminate(p)
+                recursive_kill(p)
             except ProcessLookupError:  # pragma: no cover
                 pass
 

--- a/loky/process_executor.py
+++ b/loky/process_executor.py
@@ -84,7 +84,7 @@ from .backend import get_context
 from .backend.context import cpu_count
 from .backend.queues import Queue, SimpleQueue
 from .backend.reduction import set_loky_pickler, get_loky_pickler_name
-from .backend.utils import recursive_kill, get_exitcodes_terminated_worker
+from .backend.utils import kill_process_tree, get_exitcodes_terminated_worker
 from .initializers import _prepare_initializer
 
 
@@ -763,7 +763,7 @@ class _ExecutorManagerThread(threading.Thread):
             _, p = self.processes.popitem()
             mp.util.debug(f"terminate process {p.name}, reason: {reason}")
             try:
-                recursive_kill(p)
+                kill_process_tree(p)
             except ProcessLookupError:  # pragma: no cover
                 pass
 

--- a/tests/_executor_mixin.py
+++ b/tests/_executor_mixin.py
@@ -4,6 +4,7 @@ import math
 import psutil
 import pytest
 import threading
+from time import sleep
 
 from loky import TimeoutError
 from loky import get_reusable_executor
@@ -143,18 +144,39 @@ class ExecutorMixin:
         # Make sure executor is not broken if it should not be
         executor = getattr(self, 'executor', None)
         if executor is not None:
-            expect_broken_pool = hasattr(method, "broken_pool")  # old pytest
-            for mark in getattr(method, "pytestmark", []):
-                if mark.name == "broken_pool":
-                    expect_broken_pool = True
-            is_actually_broken = executor._flags.broken is not None
-            assert is_actually_broken == expect_broken_pool
+            try:
+                # old pytest markers:
+                expect_broken_pool = hasattr(method, "broken_pool")
+                # new pytest markers:
+                for mark in getattr(method, "pytestmark", []):
+                    if mark.name == "broken_pool":
+                        expect_broken_pool = True
 
-            t_start = time.time()
-            executor.shutdown(wait=True, kill_workers=True)
-            dt = time.time() - t_start
-            assert dt < 10, "Executor took too long to shutdown"
-        _check_subprocesses_number(executor, 0)
+                is_actually_broken = executor._flags.broken is not None
+                if expect_broken_pool:
+                    for _ in range(10):
+                        # The executor manager thread can take some time to
+                        # mark the executor broken.
+                        is_actually_broken = executor._flags.broken is not None
+                        if is_actually_broken:
+                            break
+                        sleep(0.1)
+                    else:
+                        raise AssertionError(
+                            "The executor was not flagged broken at the end of "
+                            f" {method.__qualname__} as expected."
+                        )
+                else:
+                    # Check that the executor is not broken right away to avoid
+                    # wasting CI time. False negative should be very rare.
+                    assert not is_actually_broken
+            finally:
+                # Always shutdown the executor, broken or not.
+                t_start = time.time()
+                executor.shutdown(wait=True, kill_workers=True)
+                dt = time.time() - t_start
+                assert dt < 10, "Executor took too long to shutdown"
+                _check_subprocesses_number(executor, 0)
 
     def _prime_executor(self):
         # Make sure that the executor is ready to do work before running the

--- a/tests/_executor_mixin.py
+++ b/tests/_executor_mixin.py
@@ -152,7 +152,6 @@ class ExecutorMixin:
                     if mark.name == "broken_pool":
                         expect_broken_pool = True
 
-                is_actually_broken = executor._flags.broken is not None
                 if expect_broken_pool:
                     for _ in range(10):
                         # The executor manager thread can take some time to
@@ -169,6 +168,7 @@ class ExecutorMixin:
                 else:
                     # Check that the executor is not broken right away to avoid
                     # wasting CI time. False negative should be very rare.
+                    is_actually_broken = executor._flags.broken is not None
                     assert not is_actually_broken
             finally:
                 # Always shutdown the executor, broken or not.

--- a/tests/_executor_mixin.py
+++ b/tests/_executor_mixin.py
@@ -171,9 +171,11 @@ class ExecutorMixin:
                     is_actually_broken = executor._flags.broken is not None
                     assert not is_actually_broken
             finally:
-                # Always shutdown the executor, broken or not.
+                # Always shutdown the executor, with SIGKILL if the executor
+                # is actually broken.
+                kill_workers = executor._flags.broken is not None
                 t_start = time.time()
-                executor.shutdown(wait=True, kill_workers=True)
+                executor.shutdown(wait=True, kill_workers=kill_workers)
                 dt = time.time() - t_start
                 assert dt < 10, "Executor took too long to shutdown"
                 _check_subprocesses_number(executor, 0)

--- a/tests/_executor_mixin.py
+++ b/tests/_executor_mixin.py
@@ -1,7 +1,6 @@
 import sys
 import time
 import math
-import psutil
 import pytest
 import threading
 from time import sleep
@@ -9,6 +8,14 @@ from time import sleep
 from loky import TimeoutError
 from loky import get_reusable_executor
 from loky.backend import get_context
+
+
+try:
+    import psutil
+    psutil_exceptions = (psutil.NoSuchProcess, psutil.AccessDenied)
+except ImportError:
+    psutil = None
+    psutil_exceptions = ()
 
 
 # Set a large timeout as it should only be reached in case of deadlocks
@@ -37,7 +44,7 @@ def _direct_children_with_cmdline(p):
                 # when the process is being terminated by the OS.
                 continue
             children_with_cmdline.append((c, cmdline))
-        except (OSError, psutil.NoSuchProcess, psutil.AccessDenied):
+        except (OSError,) + psutil_exceptions:
             # These errors indicate that the process has terminated while
             # we were processing the info. Just discard it.
             pass
@@ -60,6 +67,10 @@ def _running_children_with_cmdline(p):
 
 def _check_subprocesses_number(executor, expected_process_number=None,
                                expected_max_process_number=None, patience=100):
+    if not psutil:
+        # psutil is not installed, we cannot check the number of subprocesses
+        return
+
     # Wait for terminating processes to disappear
     children_cmdlines = _running_children_with_cmdline(psutil.Process())
     pids_cmdlines = [(c.pid, cmdline) for c, cmdline in children_cmdlines]
@@ -190,6 +201,9 @@ class ExecutorMixin:
 
     @classmethod
     def check_no_running_workers(cls, patience=5, sleep_duration=0.01):
+        if psutil is None:
+            return
+
         deadline = time.time() + patience
 
         while time.time() <= deadline:

--- a/tests/_test_process_executor.py
+++ b/tests/_test_process_executor.py
@@ -894,6 +894,7 @@ class ExecutorTest:
             self.executor.submit(id, data).result()
 
     def test_memory_leak_protection(self):
+        pytest.importorskip("psutil")  # cannot work without psutil
         self.executor.shutdown(wait=True)
 
         executor = self.executor_type(1, context=self.context)

--- a/tests/_test_process_executor.py
+++ b/tests/_test_process_executor.py
@@ -395,7 +395,7 @@ class ExecutorShutdownTest:
         assert stdout.strip() == "apple"
 
     @classmethod
-    def _test_recursive_kill(cls, depth):
+    def _test_shutdown_and_kill_workers(cls, depth):
         executor = cls.executor_type(
             max_workers=2, context=cls.context,
             initializer=_executor_mixin.initializer_event,
@@ -407,11 +407,11 @@ class ExecutorShutdownTest:
             executor.submit(sleep_and_return, 30, 42).result()
             executor.shutdown()
         else:
-            f = executor.submit(cls._test_recursive_kill, depth + 1)
+            f = executor.submit(cls._test_shutdown_and_kill_workers, depth + 1)
             f.result()
 
-    def test_recursive_kill(self):
-        f = self.executor.submit(self._test_recursive_kill, 1)
+    def test_shutdown_and_kill_workers(self):
+        f = self.executor.submit(self._test_shutdown_and_kill_workers, 1)
         # Wait for the nested executors to be started
         _executor_mixin._test_event.wait()
 

--- a/tests/test_loky_backend.py
+++ b/tests/test_loky_backend.py
@@ -1,11 +1,9 @@
 import os
 import sys
 import time
-import psutil
 import pytest
 import signal
 import pickle
-import platform
 import socket
 import multiprocessing as mp
 from multiprocessing.connection import wait
@@ -694,6 +692,7 @@ def _run_nested_delayed(depth, delay, event):
 
 @pytest.mark.parametrize("use_psutil", [True, False])
 def test_kill_process_tree(use_psutil):
+    psutil = pytest.importorskip("psutil")
     event = ctx_loky.Event()
     p = ctx_loky.Process(target=_run_nested_delayed, args=(4, 1000, event))
     p.start()

--- a/tests/test_loky_backend.py
+++ b/tests/test_loky_backend.py
@@ -13,7 +13,7 @@ from tempfile import mkstemp
 
 from loky.backend import get_context
 from loky.backend.context import START_METHODS
-from loky.backend.utils import recursive_kill
+from loky.backend.utils import kill_process_tree
 
 from .utils import TimingWrapper, check_subprocess_call
 from .utils import with_parallel_sum, _run_openmp_parallel_sum
@@ -66,7 +66,7 @@ class TestLokyBackend:
         """Clean up the test environment from any remaining subprocesses.
         """
         for child_process in cls.active_children():
-            recursive_kill(child_process)
+            kill_process_tree(child_process)
 
     def test_current(self):
 
@@ -693,19 +693,19 @@ def _run_nested_delayed(depth, delay, event):
 
 
 @pytest.mark.parametrize("use_psutil", [True, False])
-def test_recursive_kill(use_psutil):
+def test_kill_process_tree(use_psutil):
     event = ctx_loky.Event()
     p = ctx_loky.Process(target=_run_nested_delayed, args=(4, 1000, event))
     p.start()
 
     # Wait for all the processes to be launched
     if not event.wait(30):
-        recursive_kill(p, use_psutil=use_psutil)
-        raise RuntimeError("test_recursive_kill was not able to launch "
+        kill_process_tree(p, use_psutil=use_psutil)
+        raise RuntimeError("test_kill_process_tree was not able to launch "
                            "all nested processes.")
 
     children = psutil.Process(pid=p.pid).children(recursive=True)
-    recursive_kill(p, use_psutil=use_psutil)
+    kill_process_tree(p, use_psutil=use_psutil)
 
     # The process can take some time finishing so we should wait up to 5s
     gone, alive = psutil.wait_procs(children, timeout=5)

--- a/tests/test_loky_backend.py
+++ b/tests/test_loky_backend.py
@@ -13,7 +13,7 @@ from tempfile import mkstemp
 
 from loky.backend import get_context
 from loky.backend.context import START_METHODS
-from loky.backend.utils import recursive_terminate
+from loky.backend.utils import recursive_kill
 
 from .utils import TimingWrapper, check_subprocess_call
 from .utils import with_parallel_sum, _run_openmp_parallel_sum
@@ -66,7 +66,7 @@ class TestLokyBackend:
         """Clean up the test environment from any remaining subprocesses.
         """
         for child_process in cls.active_children():
-            recursive_terminate(child_process)
+            recursive_kill(child_process)
 
     def test_current(self):
 
@@ -693,19 +693,19 @@ def _run_nested_delayed(depth, delay, event):
 
 
 @pytest.mark.parametrize("use_psutil", [True, False])
-def test_recursive_terminate(use_psutil):
+def test_recursive_kill(use_psutil):
     event = ctx_loky.Event()
     p = ctx_loky.Process(target=_run_nested_delayed, args=(4, 1000, event))
     p.start()
 
     # Wait for all the processes to be launched
     if not event.wait(30):
-        recursive_terminate(p, use_psutil=use_psutil)
-        raise RuntimeError("test_recursive_terminate was not able to launch "
+        recursive_kill(p, use_psutil=use_psutil)
+        raise RuntimeError("test_recursive_kill was not able to launch "
                            "all nested processes.")
 
     children = psutil.Process(pid=p.pid).children(recursive=True)
-    recursive_terminate(p, use_psutil=use_psutil)
+    recursive_kill(p, use_psutil=use_psutil)
 
     # The process can take some time finishing so we should wait up to 5s
     gone, alive = psutil.wait_procs(children, timeout=5)

--- a/tests/test_loky_module.py
+++ b/tests/test_loky_module.py
@@ -4,7 +4,6 @@ import sys
 import shutil
 import tempfile
 from subprocess import check_output
-import subprocess
 
 import pytest
 

--- a/tests/test_reusable_executor.py
+++ b/tests/test_reusable_executor.py
@@ -18,6 +18,11 @@ from loky.process_executor import BrokenProcessPool, ShutdownExecutorError
 from loky.reusable_executor import _ReusablePoolExecutor
 import cloudpickle
 
+try:
+    import psutil
+except ImportError:
+    psutil = None
+
 from ._executor_mixin import ReusableExecutorMixin
 from .utils import TimingWrapper, id_sleep, check_python_subprocess_call
 from .utils import filter_match
@@ -79,7 +84,6 @@ def c_exit(exitcode=0):
 
 def sleep_then_check_pids_exist(arg):
     """Sleep for some time and the check if all the passed pids exist"""
-    psutil = pytest.importorskip("psutil")
     time, pids = arg
     sleep(time)
     res = True
@@ -90,7 +94,6 @@ def sleep_then_check_pids_exist(arg):
 
 def kill_friend(pid, delay=0):
     """Function that send SIGKILL at process pid"""
-    psutil = pytest.importorskip("psutil")
     sleep(delay)
     try:
         os.kill(pid, SIGKILL)
@@ -340,6 +343,7 @@ class TestExecutorDeadLock(ReusableExecutorMixin):
     @pytest.mark.parametrize("n_proc", [1, 2, 5, 13])
     def test_crash_races(self, n_proc):
         """Test the race conditions in reusable_executor crash handling"""
+        pytest.importorskip("psutil")  # required for kill_friend & co
 
         if (sys.platform == 'win32' and sys.version_info >= (3, 8)
                 and n_proc > 5):
@@ -438,6 +442,8 @@ class TestTerminateExecutor(ReusableExecutorMixin):
 
     def test_shutdown_deadlock(self):
         """Test recovery if killed after resize call"""
+        pytest.importorskip("psutil")  # required for kill_friend & co
+
         # Test the executor.shutdown call do not cause deadlock
         executor = get_reusable_executor(max_workers=2, timeout=None)
         executor.map(id, range(2))  # start the worker processes
@@ -483,6 +489,8 @@ class TestTerminateExecutor(ReusableExecutorMixin):
 class TestResizeExecutor(ReusableExecutorMixin):
     def test_reusable_executor_resize(self):
         """Test reusable_executor resizing"""
+        # required by sleep_then_check_pids_exist
+        pytest.importorskip("psutil")
 
         executor = get_reusable_executor(max_workers=2, timeout=None)
         executor.map(id, range(2))
@@ -540,6 +548,8 @@ class TestResizeExecutor(ReusableExecutorMixin):
 
     def test_kill_after_resize_call(self):
         """Test recovery if killed after resize call"""
+        pytest.importorskip("psutil")  # required for kill_friend & co
+
         # Test the executor resizing called before a kill arrive
         executor = get_reusable_executor(max_workers=2, timeout=None)
         executor.map(id, range(2))  # trigger the creation of worker processes

--- a/tests/test_reusable_executor.py
+++ b/tests/test_reusable_executor.py
@@ -2,7 +2,6 @@ import os
 import sys
 import gc
 import ctypes
-import psutil
 import pytest
 import warnings
 import threading
@@ -80,6 +79,7 @@ def c_exit(exitcode=0):
 
 def sleep_then_check_pids_exist(arg):
     """Sleep for some time and the check if all the passed pids exist"""
+    psutil = pytest.importorskip("psutil")
     time, pids = arg
     sleep(time)
     res = True
@@ -90,6 +90,7 @@ def sleep_then_check_pids_exist(arg):
 
 def kill_friend(pid, delay=0):
     """Function that send SIGKILL at process pid"""
+    psutil = pytest.importorskip("psutil")
     sleep(delay)
     try:
         os.kill(pid, SIGKILL)

--- a/tox.ini
+++ b/tox.ini
@@ -6,10 +6,12 @@ skip_missing_interpreters=True
 [testenv]
 passenv = NUMBER_OF_PROCESSORS LOKY_MAX_CPU_COUNT
 usedevelop = True
+# Do not install psutil on Python 3.8 to have some CI runs that
+# tests that loky has no hard dependency on psutil.
 deps =
      pytest
      pytest-timeout
-     psutil
+     psutil ; python_version != '3.8'
      coverage
      viztracer ; python_version >= '3.8' and python_version <= '3.10'
      numpy ; implementation_name == 'cpython'

